plugins/contactpressuresensorsarray/src/ContactPressureSensorsArrayDriver.cpp

```diff
@@ -0,0 +1,516 @@
+    // Consider W: World Frame, L: Link Origin Frame, C: Link CG Frame
+  
+    
+#if GAZEBO_MAJOR_VERSION >= 8
+    // Get Rotation and position of CG frame with respect to the world 
+    // Gives W_R_C and W_o_C
+    ignition::math::Pose3d W_H_C = this->m_sensorLink->WorldInertialPose();
+    
+    // Get position of CG with respect to world, with orientation same as link origin frame
+    // Gives W_R_L and W_o_C
+    ignition::math::Pose3d W_H_C_rotL = this->m_sensorLink->WorldCoGPose();
+  
+    // Get absolute pose of the link
+    ignition::math::Pose3d W_H_L = this->m_sensorLink->WorldPose();
+#else
+    gazebo::math::Pose tmp = this->m_sensorLink->GetWorldInertialPose();
+    ignition::math::Pose3d W_H_C = ignition::math::Pose3d(tmp.pos.x, tmp.pos.y, tmp.pos.z, tmp.rot.w, tmp.rot.x, tmp.rot.y, tmp.rot.z);
+    
+    tmp = this->m_sensorLink->GetWorldCoGPose();
+    ignition::math::Pose3d W_H_C_rotL = ignition::math::Pose3d(tmp.pos.x, tmp.pos.y, tmp.pos.z, tmp.rot.w, tmp.rot.x, tmp.rot.y, tmp.rot.z);
+    
+    tmp = this->m_sensorLink->GetWorldPose();
+    ignition::math::Pose3d W_H_L = ignition::math::Pose3d(tmp.pos.x, tmp.pos.y, tmp.pos.z, tmp.rot.w, tmp.rot.x, tmp.rot.y, tmp.rot.z);
+# endif   
+    
+
+#if DEBUG  
+    std::cout << " Absolute pose [ " << W_H_L.Pos().X() << " " <<  W_H_L.Pos().Y() << " " << W_H_L.Pos().Z() << " ]"<< std::endl;
+#endif
+  
+    // To find L_R_C and L_o_C
+    // L_R_C = L_R_W * W_R_C
+    // L_o_C = L_R_W (W_o_C - W_o_L)
+    ignition::math::Matrix3d L_R_W = GetQuaternionAsMatrix3d(W_H_C_rotL.Rot().Inverse());
+    this->m_linkOrigin_Rot_linkCG = L_R_W*(GetQuaternionAsMatrix3d(W_H_C.Rot()));
+    this->m_linkOrigin_Pos_linkCG = L_R_W*(W_H_C.Pos() - W_H_L.Pos());
+
+#if DEBUG  
+     std::cout << " Inertial COG pose [ " << W_H_C.Pos().X() << " " <<  W_H_C.Pos().Y() << " " << W_H_C.Pos().Z() << " ]"<< std::endl;
+  
+     std::cout << "L_o_C: \n [ " << m_linkOrigin_Pos_linkCG[0] << " " <<  m_linkOrigin_Pos_linkCG[1] << " " << m_linkOrigin_Pos_linkCG[2] << " ]"<< std::endl;
+     std::cout << "L_R_C: \n [ " << m_linkOrigin_Rot_linkCG(0, 0) <<  " " << m_linkOrigin_Rot_linkCG(0, 1) << " " << m_linkOrigin_Rot_linkCG(0, 2) << std::endl <<  
+                                    m_linkOrigin_Rot_linkCG(1, 0) << " " << m_linkOrigin_Rot_linkCG(1, 1) << " " << m_linkOrigin_Rot_linkCG(1, 2)  << std::endl << 
+                                    m_linkOrigin_Rot_linkCG(2,0) << " " << m_linkOrigin_Rot_linkCG(2, 1) << " " << m_linkOrigin_Rot_linkCG(2, 2)  << std::endl;
+#endif  
+   
+    return true;
+}
+
+int GazeboYarpContactPressureSensorsArrayDriver::read(yarp::sig::Vector& out)
+{
+    yarp::os::LockGuard guard(m_dataMutex);
+  
+    // send a zero vector in case the data is not computed already
+    if (!m_dataAvailable)
+    {
+        // Introducing average time delay for the output data to be ready
+        usleep(9000);
```

Comment:
This is evil. 
If no data is available, simply return `AS_ERROR` .

Reply:
```
    if (!m_dataAvailable)
    {
        // Introducing average time delay for the output data to be ready
        usleep(9000);
        return AS_TIMEOUT;
    }
```

Is this block ok ? I'm considering the `usleep(9000)`, because the first measurement after all the computations becomes ready by an average time of 6ms, and during that time, this `read` method floods the output with timeout prints. If you think this is not necessary, I will remove the usleep as well.

Reply:
What is the period of the analog wrapper? Even if it is 1 ms, there should be just ~6 messages printed due to this. Hard-coded delays are typically problematic especially in complex systems, and personally I prefer to avoid them as much as possible.

Reply:
I understand the rationale. The period was set to a very low value. I agree that it's better to play with higher level parameters rather than introducing hard coded delays. Thank you!